operator/builtin/transformer/retain/retain.go

```diff
@@ -0,0 +1,116 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retain
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-log-collection/entry"
+	"github.com/open-telemetry/opentelemetry-log-collection/operator"
+	"github.com/open-telemetry/opentelemetry-log-collection/operator/helper"
+)
+
+func init() {
+	operator.Register("retain", func() operator.Builder { return NewRetainOperatorConfig("") })
+}
+
+// NewRetainOperatorConfig creates a new retain operator config with default values
+func NewRetainOperatorConfig(operatorID string) *RetainOperatorConfig {
+	return &RetainOperatorConfig{
+		TransformerConfig: helper.NewTransformerConfig(operatorID, "retain"),
+	}
+}
+
+// RetainOperatorConfig is the configuration of a retain operator
+type RetainOperatorConfig struct {
+	helper.TransformerConfig `mapstructure:",squash" yaml:",inline"`
+	Fields                   []entry.Field `mapstructure:"fields" json:"fields" yaml:"fields"`
+}
+
+// Build will build a retain operator from the supplied configuration
+func (c RetainOperatorConfig) Build(context operator.BuildContext) ([]operator.Operator, error) {
+	transformerOperator, err := c.TransformerConfig.Build(context)
+	if err != nil {
+		return nil, err
+	}
+	if c.Fields == nil {
+		return nil, fmt.Errorf("retain: 'fields' is empty")
+	}
+
+	retainOp := &RetainOperator{
+		TransformerOperator: transformerOperator,
+		Fields:              c.Fields,
+	}
+
+	for _, field := range c.Fields {
+		typeCheck := field.String()
+		if strings.HasPrefix(typeCheck, "$resource") {
+			retainOp.ContainsResourceFields = true
+			continue
+		}
+		if strings.HasPrefix(typeCheck, "$attributes") {
+			retainOp.ContainsAttributeFields = true
+			continue
+		}
+		retainOp.ContainsBodyFields = true
+	}
+	return []operator.Operator{retainOp}, nil
+}
+
+// RetainOperator keeps the given fields and deletes the rest.
+type RetainOperator struct {
+	helper.TransformerOperator
+	Fields                  []entry.Field
+	ContainsBodyFields      bool
+	ContainsAttributeFields bool
+	ContainsResourceFields  bool
+}
+
+// Process will process an entry with a retain transformation.
+func (p *RetainOperator) Process(ctx context.Context, entry *entry.Entry) error {
+	return p.ProcessWith(ctx, entry, p.Transform)
+}
+
+// Transform will apply the retain operation to an entry
+func (p *RetainOperator) Transform(e *entry.Entry) error {
+	newEntry := entry.New()
+	newEntry.Timestamp = e.Timestamp
+
+	if !p.ContainsAttributeFields {
+		newEntry.Attributes = e.Attributes
+	}
+	if !p.ContainsResourceFields {
+		newEntry.Resource = e.Resource
+	}
```

Comment:
Super nit, but would prefer that handling of resource and attributes are in the same order as `Build`.